### PR TITLE
Alerting: Add support for Unified Alerting mute timings.

### PR DESCRIPTION
**Pull request by alexweav**
Adds support for mute timings. In draft, while a couple loose ends are tied up with the golang client.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -77,3 +77,5 @@ require (
 	google.golang.org/grpc v1.47.0 // indirect
 	google.golang.org/protobuf v1.28.0 // indirect
 )
+
+replace github.com/grafana/grafana-api-golang-client v0.9.0 => "../grafana-api-golang-client"
```
Comment (alexweav) [inline go.mod]: Temporary, until changes get propagated.

File: grafana/resource_playlist.go [hunk 2]
```diff
@@ -74,27 +72,28 @@ func ResourcePlaylist() *schema.Resource {
 }
 
 func CreatePlaylist(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*client).gapi
+	/*client := meta.(*client).gapi
```
Comment (alexweav) [inline grafana/resource_playlist.go]: Playlist changes are temporary due to an unrelated breaking change in the golang client.

File: grafana/resource_playlist_test.go [hunk 3]
```diff
@@ -177,7 +177,7 @@ func testAccPlaylistConfigBasic(name string) string {
 resource "grafana_playlist" "test" {
 	name     = %[1]q
 	interval = "5m"
-	
+
```
Comment (alexweav) [inline grafana/resource_playlist_test.go:180]: Simple whitespace inconsistency that my editor kept wanting to fix.

File: grafana/resource_alerting_mute_timing.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The name of the mute timing.",
+			},
+
+			"intervals": {
+				Type:        schema.TypeList,
```
Comment (alexweav) [inline grafana/resource_alerting_mute_timing.go:43]: A list rather than a set here is a fairly nuanced decision. We need to reply on diff-suppression deeper in this file, but turns out that diff-suppression is completely broken in Terraform if there is a set anywhere on the path to that field.

This happens because TF represents sets inside the state object using the set's hash as a key, like:
```
intervals: map[string]interface{} {
    "hash-of-elem-1": <elem-1-data>,
    "hash-of-elem-2": <elem-2-data>,
}
```

Since the hash is the key, a change on any field no matter how deeply nested will cause the hash to be different. Terraform will think that this is a deletion and creation of the member at every layer, not an update.

When using diff-suppression, the arguments to the diff suppression hook end up being completely useless, it's all either empty strings or numbers representing the set's cardinality (weird, right?) This core reason is also coincidentally why the output of `tf plan` is so strange on sets, it shows every element as being changed rather than the one that actually got changed. Interesting issue! We just switch to lists to work around it.

Comment (JohnnyQQQQ) [reply grafana/resource_alerting_mute_timing.go:43]: Nit: should we add this as a comment above the field?

File: grafana/resource_alerting_mute_timing.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	}
+	return mt
+}
+
+func packIntervals(nts []gapi.TimeInterval) []interface{} {
+	if nts == nil {
+		return nil
+	}
+
+	intervals := make([]interface{}, 0)
```
Comment (JohnnyQQQQ) [inline grafana/resource_alerting_mute_timing.go]:
```suggestion
	intervals := make([]interface{}, len(nts))
```

File: grafana/resource_alerting_mute_timing.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+	return map[string]string{
+		"start": time.StartMinute,
+		"end":   time.EndMinute,
+	}
+}
+
+func unpackTimeRange(raw interface{}) gapi.TimeRange {
+	vals := raw.(map[string]interface{})
+	return gapi.TimeRange{
+		StartMinute: vals["start"].(string),
```
Comment (JohnnyQQQQ) [inline grafana/resource_alerting_mute_timing.go:308]: Do we have a guarantee that this will be a string? I assume yes, as the type validation should run prior as part of the runtime.

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM, added a few comments/questions.

File: grafana/resource_alerting_mute_timing.go (hunk 5 above)
Comment (alexweav) [reply grafana/resource_alerting_mute_timing.go]: Doesn't quite work because we use `append`. The array ends up 2x as long as expected with the first half all empty structs. Instead I did `make([]interface{}, 0, len(nts))` which has the desired effect, size is 0 but len(nts) element preallocated.

File: grafana/resource_alerting_mute_timing.go (hunk 6 above)
Comment (alexweav) [reply grafana/resource_alerting_mute_timing.go:308]: Yes, it's the same type as whatever we packed it into the state as. So the `pack` function directly above this one sort of mirrors it, and packs the other way around from `string` to `interface{}`.

This is pretty common practice in the provider, you'll see lots of type casting among all the types when reading from the data object. It's all over the place in nearly every file.
